New/fql

Deepak-Vohra opened the pull request:
Added a page on Full Query Logging, a new feature.

Deepak-Vohra committed:
Merge pull request #2 from apache/trunk

Sync Fork January 6

Deepak-Vohra committed:
Add files via upload

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+- Disk utilization is bounded by deleting old log segments once a configurable size is reached
+- The on disk serialization uses a flexible schema binary format (chronicle-wire) making it easy to skip unrecognized fields, add new ones, and omit old ones.
+- Can be enabled and configured via JMX, disabled, and reset (delete on disk data), logging path is configurable via both JMX and YAML
+- Introduce new ``fqltool`` in ``/bin`` that currently implements Dump which can dump in a readable format full query logs as well as follow active full query logs
```
The fql tool set now includes `Dump`, `Replay` and `Compare`.

Deepak-Vohra replied:
Mentioned in next paragraph.

"The audit logs also include CQL requests but full query logging is dedicated to CQL requests only with features such as FQL Replay and FQL Compare that are not available in audit logging."

yifan-c replied:
If the `dump` is mentioned as an item in the list, the other 2 should be added too.

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+
+Objective
+^^^^^^^^^^ 
+Full Query Logging logs all requests to the CQL interface. The full query logs could be used for debugging, performance benchmarking,  testing and auditing CQL queries. The audit logs also include CQL requests but full query logging is dedicated to CQL requests only with features such as FQL replay and FQL compare that are not available in audit logging. Audit logging is for auditing the database ...
```
> Audit logging is for auditing the database activity and may be improved to add auditing of other database activity beside authorization and CQL in future versions.

The introduction of `Audit logging` does not align with the topic of FQL introduction.

yifan-c replied:
> `performance benchmarking,  testing`

Remove extra space.

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+
+Full Query Logger
+^^^^^^^^^^^^^^^^^^ 
+The Full Query Logger is a logger that logs entire query contents after the query finishes (or times out). Queries are logged in one of two modes: single query or batch of queries. The log for an invocation of a batch of queries includes the following attributes:
```
FQL only logs the queries that successfully complete. The other queries (e.g. timed out, failed) are not to be logged.

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+ queryState - Timestamp state associated with the query invocation
+ batchTimeMillis - Approximate time in milliseconds since the epoch since the batch was invoked
+
+Bin log is a  quick and dirty binary log that is kind of a NIH version of binary logging with a traditional logging framework. It's goal is good enough performance, predictable footprint, simplicity in terms of implementation and configuration and most importantly minimal impact on producers of log records. Performance safety is accomplished by feeding items to the binary log using a weighted que ...
```
`NIH` most likely stands for `not invented here`. However, it does not help to understand, although the abbreviation exist in the source code comment. I would suggest to just drop the clause.

`Bin log` was only used once in the entire document. How is it associate with FQL? 

`It's goal is` should be `Its goal is` or just `The goal is`

Probably have the following

> Full query logging is backed up by `BinLog`. `BinLog` is a quick and dirty binary log. Its goal is...

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+
+Weighted queue is a wrapper around any blocking queue that turns it into a blocking weighted queue. The queue will weigh each element being added and removed. Adding to the queue is blocked if adding would violate the weight bound. If an element weighs in at larger than the capacity of the queue then exactly one such element will be allowed into the queue at a time. If the weight of an object cha ...
+
+The log for single CQL query includes the following attributes:
```
This should be sitting together with the one regarding batch query attributes above.

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+
+Bin log is a  quick and dirty binary log that is kind of a NIH version of binary logging with a traditional logging framework. It's goal is good enough performance, predictable footprint, simplicity in terms of implementation and configuration and most importantly minimal impact on producers of log records. Performance safety is accomplished by feeding items to the binary log using a weighted que ...
+
+Weighted queue is a wrapper around any blocking queue that turns it into a blocking weighted queue. The queue will weigh each element being added and removed. Adding to the queue is blocked if adding would violate the weight bound. If an element weighs in at larger than the capacity of the queue then exactly one such element will be allowed into the queue at a time. If the weight of an object cha ...
```
The introduction of the weighted queue is too detailed (especially the exceptions, the unit tests, etc.) that I feel only Cassandra developer will be interested in. Could we have a high-level introduction about the weighted queue, and move the details into a quoted note section for developers?

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+ query CQL query text
+ queryOptions Options associated with the query invocation
+ queryState Timestamp state associated with the query invocation
+ queryTimeMillis Approximate time in milliseconds since the epoch since the batch was invoked
```
The format is not consistent with batch query attributes. Please unify
 
`FieldName - Description`

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+::
+
+ sudo mkdir -p /tmp/cassandrafullquerylog 
+ sudo chmod -R 777 /tmp/cassandrafullquerylog 
```
Why `777`? No files are expected to be executed.

Deepak-Vohra committed:
Update fqllogging.rst

Deepak-Vohra committed:
Update fqllogging.rst

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,2096 @@
+      Query: SELECT * FROM system.peers
+      Values: 
+      
+      Type: single-query
```
Can we truncate the entries? It makes the page unnecessarily long.

Deepak-Vohra committed:
Update fqllogging.rst

Deepak-Vohra committed:
Update fqllogging.rst

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,705 @@
+- The files in the chronicle queue that have already rolled
+- The number of bytes in store files that have already rolled
+
+FQL logger sequence is as follows;
```
`follows;` -> `follows:`

yifan-c reviewed (commented):
Appreciate the efforts on making those improvements! 
If possible, please further reduce the dumped content from the cq4 file.

Deepak-Vohra committed:
Update fqllogging.rst

Deepak-Vohra commented:
- Removed a few more duplicate entries.
-  As the page is on FQL, best to list what all FQL provides even if takes slightly more page. 
- Added a comment that the log is truncated. 
- Made all references to "as follows" consistent by adding : suffix.

yifan-c commented on doc/source/new/fqllogging.rst:
```diff
@@ -0,0 +1,689 @@
+ queryState - Timestamp state associated with the query invocation
+ batchTimeMillis - Approximate time in milliseconds since the epoch since the batch was invoked
+ 
+ The log for single CQL query includes the following attributes:
```
Please remove the heading spaces so this line is not included in the code block

Deepak-Vohra committed:
Update fqllogging.rst

dineshjoshi commented:
This has been committed.
